Safely interrupt the GC finalize thread on shutdown [Case1363998]

bholmes opened the pull request:
During shutdown we wait for 40 seconds to allow all finalizers to
complete. After this time the code attempts to terminate or suspend the
thread which can cause crashes or hangs.

This change adds flags to the finalizer thread to break long running
loops and allow the thread to safely exit.

Depends on bdwgc PR landing:
https://github.com/Unity-Technologies/bdwgc/pull/66


- Should this pull request have release notes?
  - [X] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [ ] Yes
  - [ ] No
- Do these changes need to be upstreamed to [mono/mono](https://github.com/mono/mono) or [dotnet/runtime](https://github.com/dotnet/runtime) repositories?
  - [ ] Yes
  - [X] No


**Release notes**

Fixed case 1363998@username:
Mono: Safely interrupt the GC finalize thread on shutdown.


**Backports**

Lets discuss below...

bholmes commented:
My last commit is a request from @TautvydasZilys to force the termination of the thread and not suspend the thread by putting it in an infinite loop. This implementation kills the finalizer thread from the main thread using the terminate and cancel functions. There is another option.. similar to the current suspend logic where the main thread queues an infinite loop for the thread to run, we could queue a call for the finalizer to call exit on itself. Is one better than the other? Are either acceptable?

TautvydasZilys commented:
> My last commit is a request from @TautvydasZilys to force the termination of the thread and not suspend the thread by putting it in an infinite loop. 

Hold on, I never mentioned killing threads :D. That is just as likely to deadlock. This is what I said:

> that thing somehow tells the finalizer thread to enter infinite loop. It's not hijacking execution, finalizer thread cooperatively enters that loop. What if instead you told it to stop processing finalizers and exit? No thread termination would be involved

And then I said we should remove the call to put the thread into an infinite loop, not replace it with TerminateThread:

> I also notice you didn't change anything with that code that puts the finalizer thread in an infinite loop. Doesn't that need to be removed or you risk it getting in that state anyway?

bholmes commented:
@TautvydasZilys There is no perfect answer here.  The first commit sets flags to exit the loops and the finalizer thread checks for the flags and exits gracefully. This change works well and I have not had it fail to exit yet. This was in response to your suggestion "What if instead you told it to stop processing finalizers and exit"

However there is still a chance that the thread will not exit and that is the case we are discussing. What to do if the thread is still running and we need to move on.  (assuming that we need to move on)

1. Doing nothing (remove the call to mono_thread_internal_suspend_for_shutdown) would allow the thread to keep running and (probably) will crash eventually when we attempt to unload the mono dll.
2. Leaving the call to mono_thread_internal_suspend_for_shutdown is the original bug. We take over the thread with a infinite loop of code running, which is mono.dll code and causes the crash when we attempt to unload mono.dll
3. My last commit terminates the thread, but could potentially deadlock the program if a lock is held
4. Similar to option 2 we could take over the thread and call exit thread. That is easy to implement, but I am not sure if that would deadlock as well, as you do not have control over what the thread is doing when the new function is called. Probably not better then option 3.

Also, full disclosure, I am not too concerned with the decision here. I feel we are solving the edge case of an edge case at this point.

joncham commented on mono/metadata/threads.c:
```diff
@@ -6016,6 +6016,12 @@ mono_thread_internal_suspend_for_shutdown (MonoInternalThread *thread)
+void
+mono_thread_internal_terminate_for_shutdown (MonoInternalThread* thread)
+{
+	mono_threads_platform_terminate_thread (thread->thread_info, 0);
```
Do we want an exit code of 0 here?

bholmes replied:
We can assume it in the Win32 impl

joncham commented on mono/utils/mono-threads-windows.c:
```diff
@@ -590,6 +590,16 @@ mono_threads_platform_exit (gsize exit_code)
+
+	handle = info->native_handle;
+	g_assert(handle);
+	TerminateThread(handle, exit_code);
```
Should we assert OS return code here for success?

bholmes replied:
> Should we assert OS return code here for success?

Dont follow

joncham replied:
```
BOOL result = TerminateThread(handle, exit_code);
g_assert (TerminateThread);
```

bholmes replied:
Fixed

bholmes commented:
Not so fast....  The Android build failed as `pthread-cancel` is not a thing.  Stay tuned.

bholmes committed:
Kill the GC finalizer thread instead of suspending it

For embedding in Unity we need to ensure the finalizer thread dies
before we unload the mono dll.

This case should be avoided with a 40 second timeout to allow pending
finalizers to finish and an attempt to "ask the thread to exit
gracefully." This is a last resort.

bholmes committed:
Safely interrupt the GC finalize thread on shutdown

During shutdown we wait for 40 seconds to allow all finalizers to
complete. After this time the code attempts to terminate or suspend the
thread which can cause crashes or hangs.

This change adds flags to the finalizer thread to break long running
loops and allow the thread to safely exit.

Fixes case 1363998
